### ISSUE TITLE
concurrent map read and map write

### ISSUE THREAD

**Issue by arthurwolf** [c9s/bbgo#399]
The error:

```
[ 'catch error in sim console log',
  '6d9bf5c4a003ce5f3ca6f45963907d2e',
  ',fatal error: concurrent map read and map write\n\ngoroutine 1 [running]:\nruntime.throw({0x121bc39, 0x0})\n\t/usr/local/go/src/runtime/panic.go:1198 +0x71 fp=0xc0012e6128 sp=0xc0012e60f8 pc=0x43c711\nruntime.mapaccess2_faststr(0x0, 0x0, {0xc00048e808, 0x7})\n\t/usr/local/go/src/runtime/map_faststr.go:116 +0x3d4 fp=0xc0012e6190 sp=0xc0012e6128 pc=0x41b6f4\ngithub.com/c9s/bbgo/pkg/bbgo.(*ExchangeSession).LastPrice(...)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/bbgo/session.go:543\ngithub.com/c9s/bbgo/pkg/strategy/grid.(*Strategy).generateGridSellOrders(0xc000441080, 0xc000503ba0)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/strategy/grid/strategy.go:137 +0x5d fp=0xc0012e6440 sp=0xc0012e6190 pc=0xd2705d\ngithub.com/c9s/bbgo/pkg/strategy/grid.(*Strategy).placeGridSellOrders(0xc000441080, {0x139e0b8, 0xc00016ca10}, 0x0)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/strategy/grid/strategy.go:317 +0x46 fp=0xc0012e64d0 sp=0xc0012e6440 pc=0xd28ba6\ngithub.com/c9s/bbgo/pkg/strategy/grid.(*Strategy).placeGridOrders(0xc000441080, {0x139e0b8, 0xc00016ca10}, 0x0)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/strategy/grid/strategy.go:367 +0x22d fp=0xc0012e65b0 sp=0xc0012e64d0 pc=0xd2918d\ngithub.com/c9s/bbgo/pkg/strategy/grid.(*Strategy).Run.func4()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/strategy/grid/strategy.go:637 +0x32f fp=0xc0012e67b8 sp=0xc0012e65b0 pc=0xd2c62f\ngithub.com/c9s/bbgo/pkg/types.(*StandardStream).EmitStart(0xc0004acd80)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/types/standardstream_callbacks.go:13 +0x3a fp=0xc0012e67e0 sp=0xc0012e67b8 pc=0x859b1a\ngithub.com/c9s/bbgo/pkg/backtest.(*Stream).Connect(0xc0004acd80, {0x7, 0x121bc5a})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/backtest/stream.go:74 +0x5e5 fp=0xc0012e6c00 sp=0xc0012e67e0 pc=0xd0eb05\ngithub.com/c9s/bbgo/pkg/bbgo.(*Environment).Connect(0xc00077e000, {0x13981d8, 0xc0004333c0})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/bbgo/environment.go:456 +0x2e5 fp=0xc0012e6de8 sp=0xc0012e6c00 pc=0xce7405\ngithub.com/c9s/bbgo/pkg/bbgo.(*Trader).Run(0xc0000de1e0, {0x13981d8, 0xc0004333c0})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/bbgo/trader.go:333 +0x32e fp=0xc0012e6f00 sp=0xc0012e6de8 pc=0xcfe6ae\ngithub.com/c9s/bbgo/pkg/cmd.glob..func2(0x1b9e8c0, {0x11f110f, 0x9, 0x9})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/cmd/backtest.go:299 +0x15eb fp=0xc0012e7d30 sp=0xc0012e6f00 pc=0xed780b\ngithub.com/spf13/cobra.(*Command).execute(0x1b9e8c0, {0xc0004b1cb0, 0x9, 0x9})\n\t/home/arthur/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x60e fp=0xc0012e7df0 sp=0xc0012e7d30 pc=0x86deae\ngithub.com/spf13/cobra.(*Command).ExecuteC(0x1b9f5e0)\n\t/home/arthur/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x3ad fp=0xc0012e7ea8 sp=0xc0012e7df0 pc=0x86e50d\ngithub.com/spf13/cobra.(*Command).Execute(...)\n\t/home/arthur/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895\ngithub.com/c9s/bbgo/pkg/cmd.Execute()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/cmd/root.go:173 +0x3e5 fp=0xc0012e7f70 sp=0xc0012e7ea8 pc=0xed2665\nmain.main()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/cmd/bbgo/main.go:8 +0x17 fp=0xc0012e7f80 sp=0xc0012e7f70 pc=0xee32f7\nruntime.main()\n\t/usr/local/go/src/runtime/proc.go:255 +0x227 fp=0xc0012e7fe0 sp=0xc0012e7f80 pc=0x43ede7\nruntime.goexit()\n\t/usr/local/go/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc0012e7fe8 sp=0xc0012e7fe0 pc=0x46e561\n\ngoroutine 11 [select]:\ndatabase/sql.(*DB).connectionOpener(0xc00049e1a0, {0x13981d8, 0xc000433400})\n\t/usr/local/go/src/database/sql/sql.go:1196 +0x93\ncreated by database/sql.OpenDB\n\t/usr/local/go/src/database/sql/sql.go:794 +0x188\n\ngoroutine 12 [select]:\ngithub.com/go-sql-driver/mysql.(*mysqlConn).startWatcher.func1()\n\t/home/arthur/go/pkg/mod/github.com/go-sql-driver/mysql@v1.5.0/connection.go:621 +0xb0\ncreated by github.com/go-sql-driver/mysql.(*mysqlConn).startWatcher\n\t/home/arthur/go/pkg/mod/github.com/go-sql-driver/mysql@v1.5.0/connection.go:618 +0x105\n\ngoroutine 13 [running]:\n\tgoroutine running on other thread; stack unavailable\ncreated by github.com/c9s/bbgo/pkg/backtest.(*Stream).Connect\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/backtest/stream.go:77 +0x6b8\n\ngoroutine 70 [runnable]:\ngithub.com/c9s/bbgo/pkg/service.(*BacktestService).scanRowsCh.func1()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/service/backtest.go:251 +0x10d\ncreated by github.com/c9s/bbgo/pkg/service.(*BacktestService).scanRowsCh\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/service/backtest.go:239 +0xbd\n,' ]
```

The config:

```
arthur@crypto:~$ cat tmp/bbgo-config/6d9bf5c4a003ce5f3ca6f45963907d2e.yaml 
---
sessions:
  binance:
    exchange: binance
    envVarPrefix: binance
backtest:
  startTime: "2020-02-19"
  endTime: "2020-02-20"
  symbols:
  - CVCUSDT
  account:
    makerCommission: 15 # TODO: Set the right values for binance, figure out what the number means
    takerCommission: 15
    balances:
      CVC: 0.0
      USDT: 10000
exchangeStrategies:
- on: binance
  grid:
    symbol: CVCUSDT
    # was: quantity: 3333.33333333333348491578
    amount: 3333.33333333333348491578
    gridNumber: 3.00000000000000000000
    profitSpread: 0.00066816000000000098
    upperPrice: 0.03559400000000000064
    lowerPrice: 0.03392359999999999820
    long: true
arthur@crypto:~$ 
```

**Comment by arthurwolf** [comment]
Same issue happened again later on:

```
[ 'catch error in sim console log',  '7c02006f6ecd379c9b6a04d11c15a4cc',
  ',fatal error: concurrent map read and map write\n\ngoroutine 1 [running]:\nruntime.throw({0x121bc39, 0x10})\n\t/usr/local/go/src/runtime/panic.go:1198 +0x71 fp=0xc0006b6128 sp=0xc0006b60f8 pc=0x43c711\nruntime.mapaccess2_faststr(0xc0006b6208, 0x4eaaed, {0xc000689e28, 0x7})\n\t/usr/local/go/src/runtime/map_faststr.go:116 +0x3d4 fp=0xc0006b6190 sp=0xc0006b6128 pc=0x41b6f4\ngithub.com/c9s/bbgo/pkg/bbgo.(*ExchangeSession).LastPrice(...)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/bbgo/session.go:543\ngithub.com/c9s/bbgo/pkg/strategy/grid.(*Strategy).generateGridSellOrders(0xc0001aef20, 0xc0002bcd00)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/strategy/grid/strategy.go:137 +0x5d fp=0xc0006b6440 sp=0xc0006b6190 pc=0xd2705d\ngithub.com/c9s/bbgo/pkg/strategy/grid.(*Strategy).placeGridSellOrders(0xc0001aef20, {0x139e0b8, 0xc000294ee0}, 0xc)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/strategy/grid/strategy.go:317 +0x46 fp=0xc0006b64d0 sp=0xc0006b6440 pc=0xd28ba6\ngithub.com/c9s/bbgo/pkg/strategy/grid.(*Strategy).placeGridOrders(0xc0001aef20, {0x139e0b8, 0xc000294ee0}, 0x7f16b6e79200)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/strategy/grid/strategy.go:367 +0x22d fp=0xc0006b65b0 sp=0xc0006b64d0 pc=0xd2918d\ngithub.com/c9s/bbgo/pkg/strategy/grid.(*Strategy).Run.func4()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/strategy/grid/strategy.go:637 +0x32f fp=0xc0006b67b8 sp=0xc0006b65b0 pc=0xd2c62f\ngithub.com/c9s/bbgo/pkg/types.(*StandardStream).EmitStart(0xc0004ced80)\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/types/standardstream_callbacks.go:13 +0x3a fp=0xc0006b67e0 sp=0xc0006b67b8 pc=0x859b1a\ngithub.com/c9s/bbgo/pkg/backtest.(*Stream).Connect(0xc0004ced80, {0x7, 0x121bc5a})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/backtest/stream.go:74 +0x5e5 fp=0xc0006b6c00 sp=0xc0006b67e0 pc=0xd0eb05\ngithub.com/c9s/bbgo/pkg/bbgo.(*Environment).Connect(0xc0002f2000, {0x13981d8, 0xc00069b040})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/bbgo/environment.go:456 +0x2e5 fp=0xc0006b6de8 sp=0xc0006b6c00 pc=0xce7405\ngithub.com/c9s/bbgo/pkg/bbgo.(*Trader).Run(0xc0000de480, {0x13981d8, 0xc00069b040})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/bbgo/trader.go:333 +0x32e fp=0xc0006b6f00 sp=0xc0006b6de8 pc=0xcfe6ae\ngithub.com/c9s/bbgo/pkg/cmd.glob..func2(0x1b9e8c0, {0x11f110f, 0x9, 0x9})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/cmd/backtest.go:299 +0x15eb fp=0xc0006b7d30 sp=0xc0006b6f00 pc=0xed780b\ngithub.com/spf13/cobra.(*Command).execute(0x1b9e8c0, {0xc0001ca3f0, 0x9, 0x9})\n\t/home/arthur/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x60e fp=0xc0006b7df0 sp=0xc0006b7d30 pc=0x86deae\ngithub.com/spf13/cobra.(*Command).ExecuteC(0x1b9f5e0)\n\t/home/arthur/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x3ad fp=0xc0006b7ea8 sp=0xc0006b7df0 pc=0x86e50d\ngithub.com/spf13/cobra.(*Command).Execute(...)\n\t/home/arthur/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895\ngithub.com/c9s/bbgo/pkg/cmd.Execute()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/cmd/root.go:173 +0x3e5 fp=0xc0006b7f70 sp=0xc0006b7ea8 pc=0xed2665\nmain.main()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/cmd/bbgo/main.go:8 +0x17 fp=0xc0006b7f80 sp=0xc0006b7f70 pc=0xee32f7\nruntime.main()\n\t/usr/local/go/src/runtime/proc.go:255 +0x227 fp=0xc0006b7fe0 sp=0xc0006b7f80 pc=0x43ede7\nruntime.goexit()\n\t/usr/local/go/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc0006b7fe8 sp=0xc0006b7fe0 pc=0x46e561\n\ngoroutine 66 [select]:\ndatabase/sql.(*DB).connectionOpener(0xc0004c09c0, {0x13981d8, 0xc00069b080})\n\t/usr/local/go/src/database/sql/sql.go:1196 +0x93\ncreated by database/sql.OpenDB\n\t/usr/local/go/src/database/sql/sql.go:794 +0x188\n\ngoroutine 67 [select]:\ngithub.com/go-sql-driver/mysql.(*mysqlConn).startWatcher.func1()\n\t/home/arthur/go/pkg/mod/github.com/go-sql-driver/mysql@v1.5.0/connection.go:621 +0xb0\ncreated by github.com/go-sql-driver/mysql.(*mysqlConn).startWatcher\n\t/home/arthur/go/pkg/mod/github.com/go-sql-driver/mysql@v1.5.0/connection.go:618 +0x105\n\ngoroutine 55 [runnable]:\ngithub.com/c9s/bbgo/pkg/types.(*StandardStream).EmitKLineClosed(0xc000877d40, {0x22e54, {0xc0001644b0, 0x7}, {0xc000164520, 0x7}, {0x0, 0xed7db5480, 0x0}, {0x0, ...}, ...})\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/types/standardstream_callbacks.go:83 +0x7a\ngithub.com/c9s/bbgo/pkg/backtest.(*Stream).Connect.func2()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/backtest/stream.go:94 +0x2a5\ncreated by github.com/c9s/bbgo/pkg/backtest.(*Stream).Connect\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/backtest/stream.go:77 +0x6b8\n\ngoroutine 82 [chan send]:\ngithub.com/c9s/bbgo/pkg/service.(*BacktestService).scanRowsCh.func1()\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/service/backtest.go:251 +0x10d\ncreated by github.com/c9s/bbgo/pkg/service.(*BacktestService).scanRowsCh\n\t/home/arthur/go/pkg/mod/github.com/c9s/bbgo@v1.20.1-0.20211223162419-49e516b80e13/pkg/service/backtest.go:239 +0xbd\n,' ]


```

**Comment by c9s** [comment]
looks like the LastPrice's map has the concurrent issue

**Comment by c9s** [comment]
fixed in 442afe8eb9baefab437a3245823af3aade89e315
